### PR TITLE
[pruner] add support for pruning BatchNorm2d

### PR DESCRIPTION
**Pull request by kazhou**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63519

If the pruner should be pruning biases along with weights, then if the model has BatchNorm2d following pruned Conv2d layers, then the corresponding channels of the BatchNorm must also be pruned.

Specifically, they need to zeroed out, rather than fully removed, since in eager mode, the dimensions between layers need to be preserved.

To do this, we add a pruning parametrization called `ZeroesParametrization` which zeroes out pruned channels, rather than removing them. We also add checks in the pruner's `prepare` and `step` method to warn the user to manually update the BatchNorm's pruned outputs with `manual_mask_update`.

Note: `manual_mask_update` and pruning BN should only be necessary if the BatchNorm's bias is non-zero.

Differential Revision: [D30349855](https://our.internmc.facebook.com/intern/diff/D30349855/)

### CODE REVIEW

**Review by raghuramank100 — COMMENTED** [review]
I am not sure I follow the user APIs when a model has batch norms? Can you elaborate on when a user needs to call manual_mask_update? I am hoping that all the information needed can be captured as part of constructing the pruner.

**Comment by kazhou** [comment]
@raghuramank100 

If `pruner.prune_bias` is True (i.e. we are pruning the entire output channel), then user API is as follows
- user must add BN layers explicitly in `config`
    - this adds the BN layers to the pruner's module groups and attaches the  `ZeroesParametrization` and BiasHook during `prepare`
- After `pruner.step` is called, the user must call `manual_mask_update(bn_layer, conv_pruned_idx)`, where `conv_pruned_idx` is the set of pruned indices of the preceding conv2d layer

Since in eager mode we can't infer which BN follows which Conv2d, we need the user to explicitly specify that.

File: test/ao/sparsity/test_pruner.py [hunk 1]
```diff
@@ -288,18 +360,25 @@ def test_step_linear(self):
...
+        if type(model) is Conv2dBN:
+            assert model.seq[1].parametrizations.weight[0].pruned_outputs == model.seq[0].parametrizations.weight[0].pruned_outputs
+            assert model.bn.parametrizations.weight[0].pruned_outputs == model.conv2d.parametrizations.weight[0].pruned_outputs
         self._check_pruner_valid_after_step(model, pruner, {1}, device)
         assert model(x).shape == (1, 64, 24, 24)
 
     def test_step_conv2d(self):
-        models = [Conv2dA(), Conv2dB(), Conv2dC()]
+        bn_model = Conv2dBN()
+        bn_config = [(bn_model.seq[0], bn_model.seq[1]), (bn_model.conv2d, bn_model.bn)]
```
Comment (vkuzo) [inline test/ao/sparsity/test_pruner.py:378]: what does this config mean?  Does it mean that `bn_model.seq[0]` and `bn_model.seq[1]` need to be pruned with the knowledge that they execute sequentially?

File: torch/ao/sparsity/experimental/pruner/base_pruner.py [hunk 2]
```diff
@@ -44,45 +51,170 @@ def _prepare(self, use_path=False, *args, **kwargs):
         self.bias_handles = []
 
         for config in self.module_groups:
+            modules = []
```
Comment (vkuzo) [inline torch/ao/sparsity/experimental/pruner/base_pruner.py:54]: what does `use_path` mean?  Maybe we could add it to the docblock?

File: torch/ao/sparsity/experimental/pruner/base_pruner.py [hunk 3]
```diff
@@ -44,45 +51,170 @@ def _prepare(self, use_path=False, *args, **kwargs):
...
             if use_path:
-                module = fqn_to_module(self.model, config['fqn'])
+                if type(config['module']) is tuple:  # (Conv2d, BN)
```
Comment (vkuzo) [inline torch/ao/sparsity/experimental/pruner/base_pruner.py:56]: Would be great if the `self.module_groups` variable was documented in the class docblock.

File: torch/ao/sparsity/experimental/pruner/base_pruner.py [hunk 4]
```diff
@@ -44,45 +51,170 @@ def _prepare(self, use_path=False, *args, **kwargs):
...
+        Note::
+            The model is modified inplace. If you need to preserve the original
+            model, use copy.deepcopy.
+
+        Args:
+        - model [nn.Module]: model to configure. The model itself is not saved
+            but used for the state_dict saving / loading.
+        - config [list]: configuration elements could either be instances of
+            nn.Module or dict maps. The dicts must have a key 'module' with the
+            value being an instance of a nn.Module.
```
Comment (vkuzo) [inline torch/ao/sparsity/experimental/pruner/base_pruner.py:115]: would be great to specify the possible formats of config here

File: torch/ao/sparsity/experimental/pruner/base_pruner.py [hunk 5]
```diff
@@ -44,45 +51,170 @@ def _prepare(self, use_path=False, *args, **kwargs):
...
+                    if type(child) in SUPPORTED_MODULES:
+                        self.config.append(child)
+                    else:
+                        if type(child) in NEEDS_ZEROS and self.prune_bias:
+                            warnings.warn(f"Models with {type(child)} layers have config provided by user.")
+                        stack.append(child)
+
+        for module_config in self.config:
+            if type(module_config) is tuple:
+                first_layer, next_layer = module_config
```
Comment (vkuzo) [inline torch/ao/sparsity/experimental/pruner/base_pruner.py:137]: nit, would be good to assert that length is two and throw a good error message if it is not two
